AoC 2018 Day4 - part1,2

Minkyu-Jeon opened the pull request.

Minkyu-Jeon committed:
AoC 2018 Day4 - part1,2

ssaifriend commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,94 @@
+                            (str/split #" ")
+                            (rest))]
+    {:time (->> (str/split time #":") (fnext) (. Integer parseInt))
+     :guard-id (if-let [guard-id (re-matches #"[0-9]+" (nth action 1))]
```
이런 경우 `(if-let ~ ~ nil)` 은 `(when-let ~ ~)` 으로 대체할 수 있읍니다!

Minkyu-Jeon replied:
아하! 뭔가 거추장스럽다 생각이 됐는데 방법이 있었군요! 감사합니다!

namenu commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,94 @@
+      minutes
+      (let [next-minutes (->> (range (:time sleep) (:time awake))
+                              (reduce (fn [acc min]
+                                        (if (nil? (get acc min))
+                                          (assoc acc min [(:guard-id sleep)])
+                                          (update acc min #(conj %1 (:guard-id sleep)))))
```
함수를 부르고자 하는 대상이 nil일 때 기본값을 제공하는 방법이 있습니다.
고차 함수라 약간 헷깔리긴 하는데, 익숙해지면 되게 편해요!

```suggestion
                                        (update acc min (fnil conj []) (:guard-id sleep)))))
```

https://clojuredocs.org/clojure.core/fnil

Minkyu-Jeon replied:
앗 이거 정말 유용한 것 같습니다! 클로저 내장함수랑 좀 더 친해져야겠네요ㅎㅎ 감사합니다!

namenu commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,94 @@
+
+;; part 2: 한 가드가 가장 오래 잠들어있던 분 * 그 가드의 ID
+(comment (let [[min [guard-id _]] (->> frequency-map
+                                       (map (fn [[k v]] [k (get-max v)]))
+                                       (vec)
```
map -> vec 은 mapv 라는 것이 있습니다.
물론 명시적으로 vector로 변환하지 않고 답을 구할 수 있으면 더 좋구요.

Minkyu-Jeon replied:
끝에 v 붙은 함수들이 그런 용도였군요!
vector로 변환하지 않고 푸는 방법도 찾아볼게요! 감사합니다 😄

namenu commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,94 @@
+                (slurp)
+                (str/split-lines)
+                (sort)
+                (map parse-line)
+                (update-guard-ids)))
```
만약 이 문제를 rescript로 풀었다면 guard-id 는 non-nil 로 타이핑하지 않았을까 싶습니다.
그렇다면 parse-line 과정에서 임시로 guard-id를 nil로 셋팅하는게 약간 어색하게 느껴지는 것 같아요.

parse-line 과 update-guard-ids 를 하나의 함수로 합치는 정도로만 바꾸어도 괜찮을 것 같고요-!
(그러면 적어도 nil인 상태의 데이터가 함수 바깥을 돌아다니진 않으니까)

Minkyu-Jeon replied:
아하 네네! 그럼 update-guard-ids 함수에서 파싱하는 함수를 파라미터로 받아 update-guard-ids 함수 내부에서 원하는 형태의 데이터를 만드는 방향으로 수정해보겠습니다!

리스크립트도 해보고 클로저도 해보니까 장단점이 느껴져서 좋네요! ㅎㅎ

wontheone1 commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,94 @@
+        (fn [acc [_ frequencies-by-minute]]
+          (->> frequencies-by-minute
+               (reduce (fn [acc2 [guard-id times]]
+                         (if (nil? (get acc2 guard-id))
```
Clojure에서는 명시적으로 불리언 값을 테스트 하지 않고 truthy falsey를 이용해서 
```
(if (get acc2 guard-id))
  then , else 브랜치 순서 바꿈)
```
식으로 하는게 더 관용적인 것 같긴 하다는 생각이 드네요~ 필수는 아닌 것 같습니다만 truthy falsey를 적극적으로 이용하면 편하다는 느낌이 듭니당.

wontheone1 commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,94 @@
+(ns aoc2018-4
+  (:require [clojure.string :as str])
+  (:require [clojure.core]))
```
```
  (:require 
    [clojure.string :as cstr]
    [clojure.core])
```
`:require` 는 한번만 쓰시고 벡터를 여러개 두시는 것이 더 관용적입니다.

Minkyu-Jeon committed:
aoc 2018 - day4 refactoring
- 네임스페이스에 require 한번만 사용하도록 수정
- 명시적 boolean 값 테스트 대신 truthy, falsey 사용하도록 수정
- fnil 함수로 nil 기본값 처리 추가
- if-let 대신 when-let 사용
- update-guard-ids, parse-line 함수 합쳐서 nil인 상태의 데이터가 밖으로 노출되지 않도록 처리

kimchhickey commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,89 @@
+                            (rest))]
+    {:time (->> (clojure.string/split time #":") (fnext) (. Integer parseInt))
+     :guard-id (when-let [guard-id (re-matches #"[0-9]+" (nth action 1))]
+                 (. Integer parseInt guard-id))}))
```
```suggestion
                 (Integer/parseInt guard-id))}))
```

kimchhickey commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,89 @@
+      new-lines)))
+
+
+(def input (->> "input/day4_input.txt"
+                (slurp)
```
```suggestion
(def input (->> (slurp "input/day4_input.txt")
```

보통 `->>` 뒤에 sequence가 오기를 기대하기 때문에 이렇게 써주는게 좋을 것 같아요.
https://stuartsierra.com/2018/07/06/threading-with-style

Minkyu-Jeon replied:
아하 이런 컨벤션이 있었군요! 감사합니다!

kimchhickey commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,89 @@
+     :guard-id (when-let [guard-id (re-matches #"[0-9]+" (nth action 1))]
+                 (. Integer parseInt guard-id))}))
+
+(defn update-guard-ids [parse-fn lines]
+  (loop [new-lines []
+         guard-id nil
+         [first & rest] (map parse-fn lines)]
+    (if first
+      (if-let [new-guard-id (:guard-id first)]
+        (recur new-lines new-guard-id rest)
+        (recur (conj new-lines (assoc first :guard-id guard-id)) guard-id rest))
+      new-lines)))
```
이거 `(map parse-fn lines)`를 순회하는 재귀함수 같은데, reduce로 고쳐보는게 좋을 것 같습니다!

Minkyu-Jeon replied:
넵! 바꿔보겠습니다! 감사합니다!

kimchhickey commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,89 @@
+                (sort)
+                (update-guard-ids parse-line)))
+
+(defn calc-sleeping-time [lines]
+  (loop [minutes {}
+         [sleep & [awake & rest]] lines]
+    (if (empty? rest)
+      minutes
+      (let [next-minutes (->> (range (:time sleep) (:time awake))
+                              (reduce (fn [acc min]
+                                        (update acc min (fnil conj []) (:guard-id sleep)))
+                                      minutes))]
+        (recur next-minutes rest)))))
```
여기도 마찬가지로 reduce로 바꿔보면 좋을 것 같아요.

loop/recur -> reduce -> map/filter + 순수함수 + reduce 

우측으로 바꿀 수 있다면 최대한 바꾸는게 좋은데, 그 이유는 역할이 분명한 함수들로 쪼개기 위함입니다.
loop/recur 보다는 reduce가 시퀀스 순회하는 로직과 초기 조건이 분리되어 보이고,
거대한 하나의 reduce보다 map/filter + 순수함수 + reduce가 더 역할이 분리된 순수함수로 나눠져있기 때문입니다.

Minkyu-Jeon replied:
네 이것도 바꿔보겠습니다! 감사합니다 🙇

kimchhickey commented on src/aoc2018_4.clj:
```diff
@@ -0,0 +1,89 @@
+        (fn [acc [_ frequencies-by-minute]]
+          (->> frequencies-by-minute
+               (reduce (fn [acc2 [guard-id times]]
+                         (if (get acc2 guard-id)
+                           (update acc2 guard-id + times)
+                           (assoc acc2 guard-id times)))
```
```suggestion
                         (update acc2 guard-id (fnil #(+ # times) 0))
```
https://clojuredocs.org/clojure.core/fnil 이런걸로 되려나여?

Minkyu-Jeon replied:
앗 요기도 fnil을 사용하면 되겠네요! 감사합니다!

Minkyu-Jeon committed:
aoc 2018 day4 refactoring

- loop/recur -> reduce로 변경
- 쓰레딩 매크로 사용 수정
- map update시 nil체크하는 부분 fnil을 사용하도록 수정
